### PR TITLE
BabelStream package

### PR DESCRIPTION
**Pull request by jessj0nes**
Measure memory transfer rates to/from global device memory on GPUs. This benchmark is similar in spirit, and based on, the STREAM benchmark for CPUs.

New package with multiple implementations of BabelStream:

- OpenCL
- CUDA
- HIP
- OpenACC
- OpenMP
- C++ Parallel STL
- Kokkos
- RAJA
- SYCL and SYCL 2020
- TBB
- Thrust (via CUDA or HIP)

### CODE REVIEW

**Comment by tldahlgren** [comment]
@rob64j @tom91136 Are you okay with being included as maintainers on this package?

**Comment by jessj0nes** [comment]
> @rob64j @tom91136 Are you okay with being included as maintainers on this package?

- Yes

**Comment by tom91136** [comment]
> @rob64j @tom91136 Are you okay with being included as maintainers on this package?

Yes

File: var/spack/repos/builtin/packages/babelstream/package.py
Comment (alalazo) [inline var/spack/repos/builtin/packages/babelstream/package.py]: One username seems wrong:
```suggestion
    maintainers = ["robj0nes ", "tomdeakin", "tom91136"]
```

**Review by alalazo — CHANGES_REQUESTED** [review]

File: var/spack/repos/builtin/packages/babelstream/package.py
Comment (alalazo) [inline var/spack/repos/builtin/packages/babelstream/package.py:55]: This might need to be revisited after we merge #29761 Can you add a TODO on top of the variant?

File: var/spack/repos/builtin/packages/babelstream/package.py
Comment (alalazo) [inline var/spack/repos/builtin/packages/babelstream/package.py:63]: Can't we use `CudaPackage` and `RocmPackage` instead of duplicating the variants?

**Review by alalazo — CHANGES_REQUESTED** [review]
A few other comments on directives

File: var/spack/repos/builtin/packages/babelstream/package.py
Comment (tomdeakin) [reply var/spack/repos/builtin/packages/babelstream/package.py:63]: Aren't these classes the whole package would need to inherit from? How do you select this based on the user choice of variant?

Comment (alalazo) [reply var/spack/repos/builtin/packages/babelstream/package.py:63]: Not sure I get what you're asking. Those two base classes are mixins, defining the `+cuda` variant etc. and a list of valid architecture for CUDA and ROCm (instead of accepting any string).

Comment (tomdeakin) [reply var/spack/repos/builtin/packages/babelstream/package.py:63]: Your Python is better than mine! Are you are suggesting we change `class Babelstream(CMakePackage)` to `class Babelstream(CMakePackage,CudaPackage,RocmPackage)`?

Does that not break the abstraction of the `model` variant here? I.e., we'd need to say `model=std-data` for that variant but spell it `+cuda` for the CUDA variant?

Further, how do we ensure that one-and-only-one variant is selected if we go via the mixins?

Comment (bernhardkaindl) [reply var/spack/repos/builtin/packages/babelstream/package.py:63]: @robj0nes can you follow up in any way? Thanks!

Comment (tomdeakin) [reply var/spack/repos/builtin/packages/babelstream/package.py:63]: I’m taking over progressing this. Is there an update to my question on the mixins?

Comment (alalazo) [reply var/spack/repos/builtin/packages/babelstream/package.py:63]: > Further, how do we ensure that one-and-only-one variant is selected if we go via the mixins?

You can use `conflicts` directives to disallow part of the search space. 

> Does that not break the abstraction of the model variant here? I.e., we'd need to say model=std-data for that variant but spell it +cuda for the CUDA variant?

I can't say much on the `model` variant, since this is package specific. It seems to me though that you can use conditional values and say the `cuda` value to `model` can be used only when `+cuda` is on:
```
        values=[
            "std-data",
            "std-indices",
            "std-ranges",
            "ocl",
            "omp",
            "hip",
            conditional("cuda", when="+cuda"),
            "kokkos",
            "sycl",
            "sycl2020",
            "acc",
            "raja",
            "tbb",
            "thrust",
        ],
```

Comment (tomdeakin) [reply var/spack/repos/builtin/packages/babelstream/package.py:63]: Thanks for the feedback.

The `model` option of the package as it is aims to provide identical functionality to the CUDA and ROCM mixins, but for *all* models. Spelling it `+cuda` but not `model=cuda` for CUDA, but `model=foo` for everything else is not sensible.

We're going to need to redesign the programming model selection to align with the mixin approach where that's supported, but provide the `+foo` notation where there is no mixin. I'm going to use the [Kokkos package](https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/kokkos/package.py) as inspiration for implementing this as there is similarity in the mechanics of choosing a back end programming model between the two packages.

Watch this space for a rewrite!

Comment (alalazo) [reply var/spack/repos/builtin/packages/babelstream/package.py:63]: What is the status of this PR?

Comment (tomdeakin) [reply var/spack/repos/builtin/packages/babelstream/package.py:63]: We are working on an update which uses the mixins. We have something working on some systems and are testing on others. We will push the update soon for your review.

**Comment by tomdeakin** [comment]
Closing in favour of #36164

**Pull request closed without merging**
